### PR TITLE
logging ems errors to rotate separate file

### PR DESCRIPTION
**Pull request by volgustleasoft**

### CODE REVIEW

File: src/Subscriber/paymentsCustomFields.php [hunk 1]
```diff
@@ -65,36 +65,44 @@ public static function getSubscribedEvents(): array
...
 
         if($idealGateway->getID() != $requestData['paymentMethodId']) {
             $customFields = ['issuers' => $idealGateway->getCustomFields()['issuers'], 'issuer_id' => ''];
         } else {
             $customFields = ['issuers' => $idealGateway->getCustomFields()['issuers'], 'issuer_id' => $requestData['emspay_issuer_id']];
         }
+        $this->updateCustomFields($event, $idealGateway->getID(), $customFields);
 
-        $this->updateIdealCustomFields($event, $idealGateway->getID(), $customFields);
+        if($afterpayGateway->getID() != $requestData['paymentMethodId']) {
```
Comment (tiutiunnykleasoft) [inline src/Subscriber/paymentsCustomFields.php]: Caught error trying to select iDeal payment if Afterpay payment method is not installed

File: src/Service/Gateway.php [hunk 2]
```diff
@@ -62,6 +62,7 @@ public function pay(
             $pre_order = $this->processOrder($transaction,$salesChannelContext);
             $order = $this->ginger->createOrder($pre_order);
         } catch (\Exception $e) {
```
Comment (tiutiunnykleasoft) [inline src/Service/Gateway.php]: There no breakpoint with print a custom message about an error while creating the order using the Afterpay payment method.
I tried to pay using Afterpay and get the 'Symphony error message'

File: src/Resources/views/storefront/component/payment/payment-fields.html.twig [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{% sw_extends '@Storefront/storefront/component/payment/payment-fields.html.twig' %}
+{% block component_payment_method_control %}
+    {{ parent() }}
+    {% if payment.description is same as('emspay_ideal') %}
+        <select id="emspay_issuer" name="emspay_issuer_id">
+            <option value="">Choose your bank</option>
```
Comment (tiutiunnykleasoft) [inline src/Resources/views/storefront/component/payment/payment-fields.html.twig:6]: If select the "Choose your bank" option, caught the unformatted error after I return to the shop using the "return to webshop" link on the payment page
(изображение)

**Review by tiutiunnykleasoft — CHANGES_REQUESTED** [review]
Issuer ID selector for iDEAL and Afterpay birthday field form and implementation work correctly.
But as I can see the implementation of error logging & catching don't finish yet.

File: src/Subscriber/paymentsCustomFields.php (hunk 1 above)
Comment (volgustleasoft) [reply src/Subscriber/paymentsCustomFields.php]: Fixed, thx.

File: src/Service/Gateway.php (hunk 2 above)
Comment (volgustleasoft) [reply src/Service/Gateway.php]: Done

File: src/Resources/views/storefront/component/payment/payment-fields.html.twig (hunk 3 above)
Comment (volgustleasoft) [reply src/Resources/views/storefront/component/payment/payment-fields.html.twig:6]: Throw new EmsPluginException when 'issuer_id' or 'birthday' are empty.

File: src/Service/Gateway.php [hunk 4]
```diff
@@ -98,8 +109,8 @@ public function finalize(
                $message ='Error during transaction';
                $message .= isset(current($order['transactions'])['reason']) ? ':'.current($order['transactions'])['reason'].'.' : '.';
                $message .= '<br> Please contact support.';
-            $this->helper->saveEMSLog($message, ['FILE' => __FILE__, 'FUNCTION' => __FUNCTION__, 'LINE' => __LINE__]);
-            print_r($message);exit;
+
+            throw new EmsPluginException($message);
             throw new CustomerCanceledAsyncPaymentException(
                 $transactionId,
                 (current($order['transactions'])['reason'])
```
Comment (tiutiunnykleasoft) [inline src/Service/Gateway.php]: Remove this Exception we don't throwing it

File: src/Service/Gateway.php [hunk 5]
```diff
@@ -57,18 +67,19 @@ public function pay(
...
             $pre_order = $this->processOrder($transaction,$salesChannelContext);
             $order = $this->ginger->createOrder($pre_order);
         } catch (\Exception $e) {
-            $this->helper->saveEMSLog($e->getMessage(), ['FILE' => __FILE__, 'FUNCTION' => __FUNCTION__, 'LINE' => __LINE__]);
-            print_r($e->getMessage()); exit;
+            throw new EmsPluginException($e->getMessage());
             throw new AsyncPaymentProcessException(
                 $transaction->getOrderTransaction()->getId(),
                 'An error occurred during the creating the EMS Online order' . PHP_EOL . $e->getMessage()
             );
```
Comment (tiutiunnykleasoft) [inline src/Service/Gateway.php]: Remove this Exception we don't throwing it

**Review by tiutiunnykleasoft — CHANGES_REQUESTED** [review]
We need to change payment status to failed for order where we throw an exception
Try this code:
$this->transactionStateHandler->fail($transaction->getOrderTransaction()->getId(), $salesChannelContext->getContext());

File: src/Service/Gateway.php (hunk 4 above)
Comment (volgustleasoft) [reply src/Service/Gateway.php]: Done

File: src/Service/Gateway.php (hunk 5 above)
Comment (volgustleasoft) [reply src/Service/Gateway.php]: Done

**Comment by volgustleasoft** [comment]
> We need to change payment status to failed for order where we throw an exception
> Try this code:
> $this->transactionStateHandler->fail($transaction->getOrderTransaction()->getId(), $salesChannelContext->getContext());

Done

**Review by tiutiunnykleasoft — APPROVED** [review]

**Pull request merged**